clone.go

```diff
@@ -0,0 +1,262 @@
+	}
+	return fragments
+}
+
+func recurseLigate(wg *sync.WaitGroup, c chan string, seedFragment Fragment, fragmentList []Fragment) {
+	// Recurse ligate simulates all possible ligations of a series of fragments. Each possible combination begins with a "seed" that fragments from the pool can be added to.
+	defer wg.Done()
+	// If the seed ligates to itself, we can call it done with a successful circularization!
+	if seedFragment.ForwardOverhang == seedFragment.ReverseOverhang {
+		c <- seedFragment.ForwardOverhang + seedFragment.Sequence
+	} else {
+		for _, newFragment := range fragmentList {
+			// If the seedFragment's reverse overhang is ligates to a fragment's forward overhang, we can ligate those together and seed another ligation reaction
+			if seedFragment.ReverseOverhang == newFragment.ForwardOverhang {
+				newSeed := Fragment{seedFragment.Sequence + seedFragment.ReverseOverhang + newFragment.Sequence, seedFragment.ForwardOverhang, newFragment.ReverseOverhang}
+				wg.Add(1)
+				go recurseLigate(wg, c, newSeed, fragmentList)
+			}
+			// This checks if we can ligate the next fragment in its reverse direction. We have to be careful though - if our seed has a palindrome, it will ligate to itself
+			// like [-> <- -> <- -> ...] infinitely. We check for that case here as well.
+			if (seedFragment.ReverseOverhang == ReverseComplement(newFragment.ReverseOverhang)) && (seedFragment.ReverseOverhang != ReverseComplement(seedFragment.ReverseOverhang)) { // If the second statement isn't there, program will crash on palindromes
+				newSeed := Fragment{seedFragment.Sequence + seedFragment.ReverseOverhang + ReverseComplement(newFragment.Sequence), seedFragment.ForwardOverhang, ReverseComplement(newFragment.ForwardOverhang)}
+				wg.Add(1)
+				go recurseLigate(wg, c, newSeed, fragmentList)
+			}
+		}
+	}
+}
+
+// Ligate simulates ligation of all possible fragment combinations into circular plasmids.
+func Ligate(fragments []Fragment, maxClones int) []CloneSequence {
+	var wg sync.WaitGroup
+	c := make(chan string, maxClones) // A buffered channel is needed to prevent blocking.
+	wg.Add(len(fragments))
+	for _, fragment := range fragments {
+		go recurseLigate(&wg, c, fragment, fragments)
+	}
+	wg.Wait()
+	close(c)
+
+	// Due to how recurseLigate works, any sequence which is used in a complete build can be used to seed
+	// a valid rotation of the complete build. This sorts those sequences out using their unique Seqhashes.
+	var outputConstructs []CloneSequence
+	var seqhashConstruct string
+	var withinConstructs bool
+	for construct := range c {
+		seqhashConstruct, _ = Seqhash(construct, "DNA", true, true)
+		withinConstructs = false
+		for _, outputConstruct := range outputConstructs {
+			if outputConstruct.Sequence == seqhashConstruct {
+				withinConstructs = true
```

Comment:
This conditional isn't being hit by tests.

Reply:
Will go over tests after initial re-review.